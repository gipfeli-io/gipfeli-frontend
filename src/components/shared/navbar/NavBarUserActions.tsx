import { IconButton, Stack, Tooltip } from '@mui/material'
import React from 'react'
import { Link } from 'react-router-dom'
import useAuth from '../../../hooks/use-auth'
import { useNavigate } from 'react-router'
import ThemeSwitcher from './ThemeSwitcher'
import LogoutIcon from '@mui/icons-material/Logout'
import MeetingRoomIcon from '@mui/icons-material/MeetingRoom'
import LoginIcon from '@mui/icons-material/Login'

const NavBarUserActions = () => {
  const { email, signOut } = useAuth()
  const navigate = useNavigate()

  if (!email) {
    return (
      <Stack spacing={2} direction={'row'}>
        <Tooltip title={'Create a free account'}>
          <IconButton color="inherit" component={Link} to="signup" id={'join-button'}>
            <MeetingRoomIcon/>
          </IconButton>
        </Tooltip>
        <Tooltip title={'Login'}>
          <IconButton color="inherit" component={Link} to="login" id={'login-button'}>
            <LoginIcon/>
          </IconButton>
        </Tooltip>
        <ThemeSwitcher/>
      </Stack>
    )
  }

  return (
    <Stack spacing={2} direction={'row'}>
      <Tooltip title={`Logout ${email}`}>
        <IconButton onClick={() => signOut(() => navigate('/login'))} color="inherit" id={'logout-button'}>
          <LogoutIcon/>
        </IconButton>
      </Tooltip>
      <ThemeSwitcher/>
    </Stack>
  )
}

export default NavBarUserActions
