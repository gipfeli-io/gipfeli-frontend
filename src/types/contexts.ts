import { Notification } from './notifications'
import { AppTheme } from './theme'
import { handleSave } from './handle-save'
import { CurrentUpload, GpxFileUpload, ImageUpload } from './media'
import { ConnectionStatus } from '../enums/connection-status'
import { ErrorInfo } from 'react'

export interface TourListContextProperties {
  deleteEvent: (id: string) => void
}

export interface ListContextProperties {
  deleteEvent: (id: string) => void
}

export interface AuthenticationContextType {
  email: string | undefined
  isAdmin: boolean
  token: string | undefined
  signIn: (email: string, password: string, callback: () => void) => Promise<void>
  signOut: (callback: () => void) => void
}

export interface NotificationContextType {
  notification: Notification | undefined
  triggerErrorNotification: (message: string) => void
  triggerSuccessNotification: (message: string) => void
  triggerOfflineNotification: () => void
  triggerSyncFailedNotification: (message: string) => void
  resetNotification: () => void
}

export interface ConnectionStatusContextType {
  isOffline: () => boolean
  updateConnectionStatus: (connectionStatus: ConnectionStatus) => void
  updateGoOnlineButtonVisibility: (isVisible: boolean) => void
  showGoOnlineButton: boolean
  isOnlineInfoBannerVisible: boolean
  updateOnlineInfoBannerVisibility: (isVisible: boolean) => void
  resetOnlineInfoBanner: () => void
}

export interface ThemeContextType {
  activeTheme: AppTheme
  toggleTheme: () => void
}

export interface ImageUploadContextType {
  files: ImageUpload[]
  save: handleSave<File[]>
  remove: (id: string) => void
  currentUploads: CurrentUpload[]
}

export interface GpxFileUploadContextType {
  file: GpxFileUpload
  save: handleSave<File>
  remove: (id: string) => void
  currentUpload?: CurrentUpload
}

export interface ErrorBoundaryContextType{
  triggerError: (error: Error, errorInfo?: ErrorInfo) => void
}
