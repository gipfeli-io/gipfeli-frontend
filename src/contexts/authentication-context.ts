import React from 'react'
import { AuthenticationContextType } from '../types/contexts'

/**
 * The context interface (AuthenticationContextType) can be found in type src/types/contexts.ts.
 *
 * The context is implemented in src/components/providers/AuthenticationProvider.tsx, which provides functionality to login/logout a user,
 * refresh an access token and also provides useful helpers like "isAdmin" which will provide you with an easy way to control access.
 *
 * We provide the hook src/hooks/use-auth, so you can easily re-use all the functionality in your components.
 */
const AuthenticationContext = React.createContext<AuthenticationContextType>(null!)

export default AuthenticationContext
