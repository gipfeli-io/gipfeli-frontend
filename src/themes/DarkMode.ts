import { createTheme } from '@mui/material/styles'
import { LocalStorageKey } from '../enums/LocalStorageKey'
import { AppTheme } from '../types/theme'

const DarkMode: AppTheme = {
  name: LocalStorageKey.DarkMode,
  theme: createTheme({
    palette: {
      mode: 'dark'
    }
  })
}

export default DarkMode
