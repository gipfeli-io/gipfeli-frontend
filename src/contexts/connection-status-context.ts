import React from 'react'
import { ConnectionStatusContextType } from '../types/contexts'

/**
 * The context interface (ConnectionStatusContextType) can be found in type src/types/contexts.ts.
 *
 * The context is implemented in src/components/providers/ConnectionStatusProvider.tsx, which provides all the functionality
 * to handle the offline - online flow in the application. There are also some helpers you can use to check if the app is offline or not.
 *
 * We provide the hook src/hooks/use-connection-status, so you can easily re-use all the functionality in your components or services.
 */
const ConnectionStatusContext = React.createContext<ConnectionStatusContextType>(null!)

export default ConnectionStatusContext
