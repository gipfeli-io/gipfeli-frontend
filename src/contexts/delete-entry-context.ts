import React from 'react'
import { DeleteEntryContextType } from '../types/contexts'

const DeleteEntryContext = React.createContext<DeleteEntryContextType>(null!)

export default DeleteEntryContext
