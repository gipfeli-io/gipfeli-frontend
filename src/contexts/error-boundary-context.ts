import React from 'react'
import { ErrorBoundaryContextType } from '../types/contexts'

// eslint-disable-next-line n/handle-callback-err
const ErrorBoundaryContext = React.createContext<ErrorBoundaryContextType>(null!)

export default ErrorBoundaryContext
