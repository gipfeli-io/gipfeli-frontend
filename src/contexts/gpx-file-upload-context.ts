import React from 'react'
import { GpxFileUploadContextType } from '../types/contexts'

const GpxFileUploadContext = React.createContext<GpxFileUploadContextType>(null!)

export default GpxFileUploadContext
