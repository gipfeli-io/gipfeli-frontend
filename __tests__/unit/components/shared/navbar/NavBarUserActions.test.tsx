import '@testing-library/jest-dom'
// @ts-ignore
import React from 'react'
// @ts-ignore
import renderer from 'react-test-renderer'
import NavBarUserActions from '../../../../../src/components/shared/navbar/NavBarUserActions'
import { MemoryRouter } from 'react-router'
import { fireEvent, render } from '@testing-library/react'
import { ThemeContextType } from '../../../../../src/types/contexts'
import DarkMode from '../../../../../src/themes/dark-mode'

const mockSignIn: jest.Mock = jest.fn()
const mockSignOut: jest.Mock = jest.fn()
const mockAuthenticationContext: any = {
  email: null,
  token: null,
  mockSignIn,
  mockSignOut
}

const mockUseLocationValue = {
  pathname: '/localhost:3000/login',
  search: '',
  hash: '',
  state: null
}

const mockThemeContext: ThemeContextType = {
  activeTheme: DarkMode,
  toggleTheme: jest.fn()
}

const mockEmail = 'test@gipfeli.io'
const mockToken = 'mockedToken'
jest.mock('../../../../../src/hooks/use-auth', () => jest.fn().mockImplementation(() => mockAuthenticationContext))
jest.mock('../../../../../src/hooks/use-theme', () => jest.fn().mockImplementation(() => mockThemeContext))

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router') as any,
  useNavigate: () => jest.fn(),
  useLocation: () => jest.fn().mockImplementation(() => mockUseLocationValue),
  useHref: () => jest.fn()
}))

describe('NavBarUserActions', () => {
  describe('test with logged in user', () => {
    beforeEach(() => {
      mockAuthenticationContext.token = mockToken
      mockAuthenticationContext.email = mockEmail
    })

    it('behaves consistently when logged in', () => {
      const tree = renderer
        .create(<MemoryRouter initialEntries={['/currentUri']}><NavBarUserActions/></MemoryRouter>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('shows logout button with name of user on hover as well as the theme switcher if logged in', async () => {
      mockAuthenticationContext.email = mockEmail
      mockAuthenticationContext.token = mockToken
      const { container, findByText } = render(
        <MemoryRouter initialEntries={['/currentUri']}><NavBarUserActions/></MemoryRouter>
      )

      const logoutButton = container.querySelector('#logout-button')
      const themeSwitcher = container.querySelector('#theme-switcher')

      expect(logoutButton).toBeInTheDocument()
      expect(themeSwitcher).toBeInTheDocument()
      fireEvent.mouseOver(logoutButton!)
      const greeting = await findByText('Logout test@gipfeli.io', { exact: false })
      expect(greeting).toBeInTheDocument()
    })
  })

  describe('test with logged out user', () => {
    beforeEach(() => {
      mockAuthenticationContext.token = null
      mockAuthenticationContext.email = null
    })

    it('behaves consistently when not logged in', () => {
      const tree = renderer
        .create(<MemoryRouter initialEntries={['/currentUri']}><NavBarUserActions/></MemoryRouter>)
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('shows join and login button and theme switcher if not logged in', async () => {
      const { container } = render(
        <MemoryRouter initialEntries={['/currentUri']}><NavBarUserActions/></MemoryRouter>
      )

      const loginButton = container.querySelector('#login-button')
      const logoutButton = container.querySelector('#logout-button')
      const themeSwitcher = container.querySelector('#theme-switcher')

      expect(loginButton).toBeInTheDocument()
      expect(themeSwitcher).toBeInTheDocument()
      expect(logoutButton).not.toBeInTheDocument()
    })
  })
})
