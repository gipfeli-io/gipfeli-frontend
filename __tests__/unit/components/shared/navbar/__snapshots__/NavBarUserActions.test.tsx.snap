// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBarUserActions test with logged in user behaves consistently when logged in 1`] = `
<div
  className="css-e53awj-MuiStack-root"
>
  <button
    aria-label="Logout test@gipfeli.io"
    aria-labelledby={null}
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
    data-mui-internal-clone-element={true}
    disabled={false}
    id="logout-button"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="LogoutIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <button
    aria-label="Toggle theme"
    aria-labelledby={null}
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
    data-mui-internal-clone-element={true}
    disabled={false}
    id="theme-switcher"
    onBlur={[Function]}
    onClick={[MockFunction]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="Brightness7Icon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6zm0-10c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`NavBarUserActions test with logged out user behaves consistently when not logged in 1`] = `
<div
  className="css-e53awj-MuiStack-root"
>
  <a
    aria-label="Create a free account"
    aria-labelledby={null}
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
    data-mui-internal-clone-element={true}
    href={[MockFunction]}
    id="join-button"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="MeetingRoomIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M14 6v15H3v-2h2V3h9v1h5v15h2v2h-4V6h-3zm-4 5v2h2v-2h-2z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </a>
  <a
    aria-label="Login"
    aria-labelledby={null}
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
    data-mui-internal-clone-element={true}
    href={[MockFunction]}
    id="login-button"
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="LoginIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M11 7 9.6 8.4l2.6 2.6H2v2h10.2l-2.6 2.6L11 17l5-5-5-5zm9 12h-8v2h8c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-8v2h8v14z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </a>
  <button
    aria-label="Toggle theme"
    aria-labelledby={null}
    className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
    data-mui-internal-clone-element={true}
    disabled={false}
    id="theme-switcher"
    onBlur={[Function]}
    onClick={[MockFunction]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="Brightness7Icon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69zM12 18c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.69 6-6 6zm0-10c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4z"
      />
    </svg>
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
